test/gtest/diag.hpp

```diff
@@ -0,0 +1,252 @@
+        }
+        else if((N != 0) && (W != 0))
+        {
+            return std::vector<size_t>({N, W});
+        }
+        else if((N != 0))
+        {
+            return std::vector<size_t>({N});
+        }
+        else
+        {
+            std::cout << "Error Input Tensor Lengths\n" << std::endl;
+            return std::vector<size_t>({0});
+        }
+    }
+
+    std::vector<size_t> ComputeStrides(std::vector<size_t> inputDim) const
+    {
+        if(!isContiguous)
+            std::swap(inputDim.front(), inputDim.back());
+        std::vector<size_t> strides(inputDim.size());
+        strides.back() = 1;
+        for(int i = inputDim.size() - 2; i >= 0; --i)
+            strides[i] = strides[i + 1] * inputDim[i + 1];
+        if(!isContiguous)
+            std::swap(strides.front(), strides.back());
+        return strides;
+    }
+};
+
+std::vector<DiagTestCase> DiagTestConfigs()
+{ // n c d h w dim
+    // clang-format off
+    return {
+        { 2,    0,   0,  0,   4,    0, true},
+        { 2,    0,   0,  0,   4,    8, true},
+        { 2,    0,   0,  0,   4,    -8, true},
+        { 16,    0,   0,  0,   16,    0, true},
+        { 16,    0,   0,  0,   16,    2, true},
+        { 16,    0,   0,  0,   16,    -2, true},
+        { 32,    0,   0,  0,    0,     0, true},
+        { 32,    0,   0,  0,    0,     2, true},
+        { 32,    0,   0,  0,    0,     -2, true},
+        { 64,    0,  0,  0,    7,     0, true},
+        { 64,    0,  0,  0,    7,     2, true},
+        { 64,    0,  0,  0,    7,     -2, true},
+        { 2,    0,   0,  0,   4,    0, false},
+        { 2,    0,   0,  0,   4,    8, false},
+        { 2,    0,   0,  0,   4,    -8, false},
+        { 16,    0,   0,  0,   16,    0, false},
+        { 16,    0,   0,  0,   16,    2, false},
+        { 16,    0,   0,  0,   16,    -2, false},
+        { 32,    0,   0,  0,    0,     0, false},
+        { 32,    0,   0,  0,    0,     2, false},
+        { 32,    0,   0,  0,    0,     -2, false},
+        { 64,    0,  0,  0,    7,     0, false},
+        { 64,    0,  0,  0,    7,     2, false},
+        { 64,    0,  0,  0,    7,     -2, false},
+      };
+    // clang-format on
+}
+
+template <typename T>
+struct DiagFwdTest : public ::testing::TestWithParam<DiagTestCase>
+{
+protected:
+    void SetUp() override
+    {
+
+        auto&& handle = get_handle();
+        diag_config   = GetParam();
+        std::cout << diag_config << std::endl;
+        auto gen_value = [](auto...) { return prng::gen_descreet_uniform_sign<T>(1e-2, 100); };
+
+        diagonal = diag_config.diagonal;
+
+        auto in_dims  = diag_config.GetInput();
+        auto inStride = diag_config.ComputeStrides(in_dims);
+
+        input = tensor<T>{in_dims, inStride}.generate(gen_value);
+
+        std::vector<size_t> out_dims;
+
+        if(input.desc.GetLengths().size() == 1)
+        {
+            size_t sz = in_dims[0] + abs(diagonal);
+            out_dims  = {sz, sz};
```

Comment:
If diagonal is != 0. out_dims is not {sz, sz}. Right?

Reply:
`sz` is calculated based on `diagonal` so `out_dims` is always {sz, sz} when the dimension of input = 1